pyroengine/core.py

```diff
@@ -0,0 +1,292 @@
+                # Open image
+                frame = Image.open(entry["frame_path"], mode="r")
+                self._alerts.append({"frame": frame, "cam_id": entry["cam_id"], "ts": entry["ts"]})
+
+    def heartbeat(self, cam_id: str) -> None:
+        """Updates last ping of device"""
+        self.api_client[cam_id].heartbeat()
+
+    def predict(self, frame: Image.Image, cam_id: Optional[str] = None) -> float:
+        """Computes the confidence that the image contains wildfire cues
+
+        Args:
+            frame: a PIL image
+            cam_id: the name of the camera that sent this image
+        Returns:
+            the predicted confidence
+        """
+
+        # Heartbeat
+        if len(self.api_client) > 0 and isinstance(cam_id, str):
+            self.heartbeat(cam_id)
+
+        # Inference with ONNX
+        pred = float(self.model(frame.convert("RGB")))
+        # Log analysis result
+        device_str = f"Camera {cam_id} - " if isinstance(cam_id, str) else ""
+        pred_str = "Wildfire detected" if pred >= self.conf_thresh else "No wildfire"
+        logging.info(f"{device_str}{pred_str} (confidence: {pred:.2%})")
+
+        # Reduce image size to save bandwidth
+        if isinstance(self.frame_size, tuple):
+            frame = frame.resize(self.frame_size, Image.BILINEAR)
+
+        # Alert
+        cam_key = cam_id or "-1"
+        if pred >= self.conf_thresh:
+            # Don't increment beyond relaxation
+            if not self.ongoing_alert[cam_key] and self.consec_dets[cam_key] < self.alert_relaxation:
+                self.consec_dets[cam_key] += 1
+
+            if self.consec_dets[cam_key] == self.alert_relaxation:
+                self.ongoing_alert[cam_key] = True
+
+            if len(self.api_client) > 0 and isinstance(cam_id, str) and self.ongoing_alert[cam_key]:
+                # Save the alert in cache to avoid connection issues
+                self._stage_alert(frame, cam_id)
+
+        # No wildfire
+        else:
+
+            if self.consec_dets[cam_key] > 0:
+                self.consec_dets[cam_key] -= 1
+            # Consider event as finished
+            if self.consec_dets[cam_key] == 0:
+                self.ongoing_alert[cam_key] = False
+
+        # Uploading pending alerts
+        if len(self._alerts) > 0:
+            self._process_alerts()
+
+        # Check if it's time to backup pending alerts
+        ts = datetime.utcnow()
+        if ts > self.last_cache_dump + timedelta(minutes=self.cache_backup_period):
+            self._dump_cache()
+            self.last_cache_dump = ts
+
+        # save frame
+        if len(self.api_client) > 0 and isinstance(self.frame_saving_period, int) and isinstance(cam_id, str):
+            self.frames_counter[cam_id] += 1
+            if self.frames_counter[cam_id] == self.frame_saving_period:
+                # Send frame to the api
+                stream = io.BytesIO()
+                frame.save(stream, format="JPEG")
```

Comment:
maybe it's not the time to do it but jpeg default compression is 75% for jpeg, it's a bit low

Reply:
Mmmh the option to downsize the image is already in place, we can check this in another PR (since we'll need some benchmarking)

Reply:
Yes I just wanted to raise the "issue", we can manage this later with a proper bench